### ISSUE TITLE
Cannot join new node

### ISSUE THREAD

**Issue by Freeaqingme** [scylladb/scylla#1231]
Scylla version (or git commit hash): 1.0.2-20160423.0732b76-ubuntu1
Cluster size: 4 nodes
OS: Linux db0 4.4.0-21-generic #37~14.04.1-Ubuntu SMP Wed Apr 20 16:33:38 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux

Howdy,

Yesterday I was able to pretty easily bootstrap a 4 node cluster.

Current cluster:
```
node0]─[~]
└──╼ nodetool status
Datacenter: irrelevant
================================
Status=Up/Down
|/ State=Normal/Leaving/Joining/Moving
--  Address          Load       Tokens  Owns    Host ID                               Rack
UN  149.X.Y.116  254.03 MB  256     ?       4a7ce2f2-4e39-4762-99c3-92a929631c6b  Z
UN  149.X.Y.117  262.92 MB  256     ?       2e5de3ab-f773-4f3b-844f-359a0931857a  Z
UN  149.X.Y.115  214.74 MB  256     ?       5698433b-7af7-4b3d-be21-dc299781d2d8  Z
UN  149.X.Y.130  248.88 MB  256     ?       d557c3be-bb0a-4e05-a545-515a53791bc0  Z
```

Now I'm trying to add a fifth node. As far as I can tell it's configured identical to the other 4 (long live Puppet). While my new node (149.X.Y.132) was joining, I see on 149.X.Y.115:
```
May  3 22:10:27 node0.company.tld scylla[10747]:  [shard 0] messaging_service - Retry verb=17 to 149.X.Y.132:0, retry=10: rpc::closed_error (connection is closed)
May  3 22:10:27 node0.company.tld scylla[10747]:  [shard 0] messaging_service - Retry verb=17 to 149.X.Y.132:0, retry=10: rpc::closed_error (connection is closed)
May  3 22:10:27 node0.company.tld scylla[10747]:  [shard 0] messaging_service - Retry verb=17 to 149.X.Y.132:0, retry=10: rpc::closed_error (connection is closed)
May  3 22:10:37 node0.company.tld scylla[10747]:  [shard 0] gossip - FatClient 149.X.Y.132 has been silent for 30000ms, removing from gossip
May  3 22:10:37 node0.company.tld scylla[10747]:  [shard 0] stream_session - stream_manager: Close all stream_session with peer = 149.X.Y.132 in on_remove
May  3 22:10:37 node0.company.tld scylla[10747]:  [shard 1] stream_session - [Stream #05fabc70-116b-11e6-9b2a-000000000002] Session with 149.X.Y.132 is complete, state=FAILED
May  3 22:10:37 node0.company.tld scylla[10747]:  [shard 1] stream_session - [Stream #05fabc70-116b-11e6-9b2a-000000000002] bytes_sent = 32133150, bytes_received = 0
May  3 22:10:37 node0.company.tld scylla[10747]:  [shard 1] stream_session - [Stream #05fabc70-116b-11e6-9b2a-000000000002] Stream failed, peers={149.X.Y.132}
```

Whereas on the new node (.132) I see:
```
May  3 22:09:34 db0 scylla[17196]:  [shard 0] storage_service - JOINING: sleeping 30000 ms for pending range setup
May  3 22:09:34 db0 scylla[17196]:  [shard 0] gossip - Node 149.X.Y.130 is now part of the cluster
May  3 22:09:34 db0 scylla[17196]:  [shard 0] gossip - InetAddress 149.X.Y.130 is now UP
May  3 22:09:34 db0 scylla[17196]:  [shard 0] storage_service - Node 149.X.Y.130 state jump to normal
May  3 22:09:34 db0 scylla[17196]:  [shard 0] compaction - Compacted 4 sstables to [/var/lib/scylla/data/system/local-7ad54392bcdd35a684174e047860b377/system-local-ka-148-Data.db:level=0, ]. 25114 bytes to 12558 (~50% of original) in 208ms = 0.0575781MB/s. ~1024 total partitions merged to 1.
May  3 22:09:34 db0 scylla[17196]:  [shard 0] gossip - do_gossip_to_live_member: Favor newly added node 149.X.Y.116
May  3 22:09:35 db0 scylla[17196]:  [shard 0] gossip - do_gossip_to_live_member: Favor newly added node 149.X.Y.117
May  3 22:09:36 db0 scylla[17196]:  [shard 0] gossip - do_gossip_to_live_member: Favor newly added node 149.X.Y.130
May  3 22:10:04 db0 scylla[17196]:  [shard 0] storage_service - JOINING: Starting to bootstrap...
May  3 22:10:04 db0 scylla[17196]:  [shard 0] stream_session - [Stream #05fabc70-116b-11e6-9b2a-000000000002] Executing streaming plan for Bootstrap
May  3 22:10:04 db0 scylla[17196]:  [shard 0] stream_session - [Stream #05fabc70-116b-11e6-9b2a-000000000002] Starting streaming to 149.X.Y.115
May  3 22:10:04 db0 scylla[17196]:  [shard 0] stream_session - [Stream #05fabc70-116b-11e6-9b2a-000000000002] Beginning stream session with 149.X.Y.115
May  3 22:10:04 db0 scylla[17196]:  [shard 0] stream_session - [Stream #05fabc70-116b-11e6-9b2a-000000000002] Starting streaming to 149.X.Y.116
May  3 22:10:04 db0 scylla[17196]:  [shard 0] stream_session - [Stream #05fabc70-116b-11e6-9b2a-000000000002] Beginning stream session with 149.X.Y.116
May  3 22:10:04 db0 scylla[17196]:  [shard 0] stream_session - [Stream #05fabc70-116b-11e6-9b2a-000000000002] Starting streaming to 149.X.Y.117
May  3 22:10:04 db0 scylla[17196]:  [shard 0] stream_session - [Stream #05fabc70-116b-11e6-9b2a-000000000002] Beginning stream session with 149.X.Y.117
May  3 22:10:04 db0 scylla[17196]:  [shard 0] stream_session - [Stream #05fabc70-116b-11e6-9b2a-000000000002] Starting streaming to 149.X.Y.130
May  3 22:10:04 db0 scylla[17196]:  [shard 0] stream_session - [Stream #05fabc70-116b-11e6-9b2a-000000000002] Beginning stream session with 149.X.Y.130
May  3 22:10:04 db0 scylla[17196]:  [shard 0] stream_session - [Stream #05fabc70-116b-11e6-9b2a-000000000002] Prepare completed with 149.X.Y.115. Receiving 3, sending 0
May  3 22:10:04 db0 scylla[17196]:  [shard 0] stream_session - [Stream #05fabc70-116b-11e6-9b2a-000000000002] Prepare completed with 149.X.Y.130. Receiving 3, sending 0
May  3 22:10:04 db0 scylla[17196]:  [shard 0] stream_session - [Stream #05fabc70-116b-11e6-9b2a-000000000002] Prepare completed with 149.X.Y.116. Receiving 3, sending 0
May  3 22:10:04 db0 scylla[17196]:  [shard 0] stream_session - [Stream #05fabc70-116b-11e6-9b2a-000000000002] Prepare completed with 149.X.Y.117. Receiving 3, sending 0
```

Also, on the new node I see:
```
nodetool status
nodetool: Scylla API server HTTP GET to URL '/snitch/datacenter' failed: Not found
See 'nodetool help' or 'nodetool help <command>'.
```
But sometimes that command just hangs, doing nothing. On the existing nodes, I temporarily (after this it returns to its 4-node state) get:

```
nodetool status
Datacenter: Irrelevant
================================
Status=Up/Down
|/ State=Normal/Leaving/Joining/Moving
--  Address          Load       Tokens  Owns    Host ID                               Rack
UN  149.X.Y.116  254.03 MB  256     ?       4a7ce2f2-4e39-4762-99c3-92a929631c6b  Z
UN  149.X.Y.117  262.92 MB  256     ?       2e5de3ab-f773-4f3b-844f-359a0931857a  Z
UN  149.X.Y.115  214.74 MB  256     ?       5698433b-7af7-4b3d-be21-dc299781d2d8  Z
UN  149.X.Y.130  248.88 MB  256     ?       d557c3be-bb0a-4e05-a545-515a53791bc0  Z
DJ  149.X.Y.132  ?          256     ?       e4f5762d-ac35-4158-b7cb-50d8e7400832  Z
```

After which it goes back to just 4 nodes.

Even though the logs may suggest there's no connection going on, shutting down scylladb on the new node, does result in these log messages on the pre-existing nodes:
```
May  3 22:17:11 node0.company.tld scylla[10747]:  [shard 0] rpc - client 149.X.Y.132: client connection dropped: read: Connection reset by peer
May  3 22:17:11 node0.company.tld scylla[10747]:  [shard 0] messaging_service - Retry verb=17 to 149.X.Y.132:0, retry=9: rpc::closed_error (connection is closed)
May  3 22:17:11 node0.company.tld scylla[10747]:  [shard 1] messaging_service - Retry verb=17 to 149.X.Y.132:0, retry=10: rpc::closed_error (connection is closed)
May  3 22:17:11 node0.company.tld scylla[10747]:  [shard 0] messaging_service - Retry verb=17 to 149.X.Y.132:0, retry=9: rpc::closed_error (connection is closed)
May  3 22:17:11 node0.company.tld scylla[10747]:  [shard 1] messaging_service - Retry verb=17 to 149.X.Y.132:0, retry=10: rpc::closed_error (connection is closed)
May  3 22:17:11 node0.company.tld scylla[10747]:  [shard 3] messaging_service - Retry verb=17 to 149.X.Y.132:0, retry=10: rpc::closed_error (connection is closed)
```

I could not really deduct anything from the logs, so tried to run it in debug mode:
```
sudo  /usr/bin/scylla --log-to-syslog 1 --log-to-stdout 0 --default-log-level info --collectd-address=127.0.0.1:25826 --collectd=1 --collectd-poll-period 3000 --network-stack posix --max-io-requests=15 --num-io-queues=3 --logger-log-level=debug
error: the argument ('debug') for option '--level' is invalid
```

Not much help there either. I'm not much of a hero when it comes to GDB, but for full brevity, this is how far I got:
```
sudo gdb /usr/bin/scylla 
GNU gdb (Ubuntu 7.7.1-0ubuntu5~14.04.2) 7.7.1
Copyright (C) 2014 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "x86_64-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
<http://www.gnu.org/software/gdb/documentation/>.
For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from /usr/bin/scylla...(no debugging symbols found)...done.
(gdb) r --log-to-syslog 1 --log-to-stdout 0 --default-log-level info --collectd-address=127.0.0.1:25826 --collectd=1 --collectd-poll-period 3000 --network-stack posix --max-io-requests=15 --num-io-queues=3
Starting program: /usr/bin/scylla --log-to-syslog 1 --log-to-stdout 0 --default-log-level info --collectd-address=127.0.0.1:25826 --collectd=1 --collectd-poll-period 3000 --network-stack posix --max-io-requests=15 --num-io-queues=3
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
EAL: Detected lcore 0 as core 0 on socket 0
EAL: Detected lcore 1 as core 0 on socket 0
EAL: Detected lcore 2 as core 0 on socket 0
EAL: Detected lcore 3 as core 0 on socket 0
EAL: Support maximum 128 logical core(s) by configuration.
EAL: Detected 4 lcore(s)
EAL: VFIO modules not all loaded, skip VFIO support...
EAL: Setting up physically contiguous memory...
[New Thread 0x7fffedf96700 (LWP 18523)]
EAL: TSC frequency is ~2294686 KHz
EAL: WARNING: cpu flags constant_tsc=yes nonstop_tsc=no -> using unreliable clock cycles !
EAL: Master lcore 0 is ready (tid=f7fda5c0;cpuset=[0])
[New Thread 0x7fffed795700 (LWP 18524)]
EAL: lcore 1 is ready (tid=ed795700;cpuset=[1])
[New Thread 0x7fffecf94700 (LWP 18525)]
EAL: lcore 2 is ready (tid=ecf94700;cpuset=[2])
[New Thread 0x7fffec793700 (LWP 18526)]
EAL: lcore 3 is ready (tid=ec793700;cpuset=[3])
[New Thread 0x7fffebbff700 (LWP 18527)]
[New Thread 0x7fffeb5ff700 (LWP 18528)]
[New Thread 0x7fffeafff700 (LWP 18529)]
[New Thread 0x7fffea9ff700 (LWP 18530)]
Scylla version 1.0.2-20160423.60307f6 starting ...

Program received signal SIG34, Real-time event 34.
[Switching to Thread 0x7fffec793700 (LWP 18526)]
0x00007ffff3d416b7 in __GI_epoll_pwait (epfd=25, events=0x7fffec78c4d0, maxevents=128, timeout=-204204361, set=0x603000111010)
    at ../sysdeps/unix/sysv/linux/epoll_pwait.c:48
48	../sysdeps/unix/sysv/linux/epoll_pwait.c: No such file or directory.
(gdb) bt
#0  0x00007ffff3d416b7 in __GI_epoll_pwait (epfd=25, events=0x7fffec78c4d0, maxevents=128, timeout=-204204361, set=0x603000111010)
    at ../sysdeps/unix/sysv/linux/epoll_pwait.c:48
#1  0x00000000004bf92b in ?? ()
#2  0x00000000004bfaa1 in ?? ()
#3  0x00000000004e0363 in ?? ()
#4  0x00000000004fa809 in ?? ()
#5  0x00000000004b162e in ?? ()
#6  0x00000000006b9395 in ?? ()
#7  0x00007ffff4014182 in start_thread (arg=0x7fffec793700) at pthread_create.c:312
#8  0x00007ffff3d4147d in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:111
(gdb) bt full
#0  0x00007ffff3d416b7 in __GI_epoll_pwait (epfd=25, events=0x7fffec78c4d0, maxevents=128, timeout=-204204361, set=0x603000111010)
    at ../sysdeps/unix/sysv/linux/epoll_pwait.c:48
        resultvar = 105759275814912
        oldtype = 0
#1  0x00000000004bf92b in ?? ()
No symbol table info available.
#2  0x00000000004bfaa1 in ?? ()
No symbol table info available.
#3  0x00000000004e0363 in ?? ()
No symbol table info available.
#4  0x00000000004fa809 in ?? ()
No symbol table info available.
#5  0x00000000004b162e in ?? ()
No symbol table info available.
#6  0x00000000006b9395 in ?? ()
No symbol table info available.
#7  0x00007ffff4014182 in start_thread (arg=0x7fffec793700) at pthread_create.c:312
        __res = <optimized out>
        pd = 0x7fffec793700
        now = <optimized out>
        unwind_buf = {cancel_jmp_buf = {{jmp_buf = {140737160754944, 5468883168902647791, 1, 0, 140737160755648, 140737160754944, 
                -5468855706036858897, -5468908959868645393}, mask_was_saved = 0}}, priv = {pad = {0x0, 0x0, 0x0, 0x0}, data = {prev = 0x0, 
              cleanup = 0x0, canceltype = 0}}}
        not_first_call = <optimized out>
        pagesize_m1 = <optimized out>
        sp = <optimized out>
        freesize = <optimized out>
        __PRETTY_FUNCTION__ = "start_thread"
#8  0x00007ffff3d4147d in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:111
No locals.
(gdb) r
The program being debugged has been started already.
Start it from the beginning? (y or n) n
Program not restarted.
(gdb) c
Continuing.

Program received signal SIG34, Real-time event 34.
[Switching to Thread 0x7fffecf94700 (LWP 18525)]
0x00007ffff3d416b7 in __GI_epoll_pwait (epfd=23, events=0x7fffecf8d4d0, maxevents=128, timeout=-204204361, set=0x602000111010)
    at ../sysdeps/unix/sysv/linux/epoll_pwait.c:48
48	in ../sysdeps/unix/sysv/linux/epoll_pwait.c
(gdb) c
Continuing.

Program received signal SIG34, Real-time event 34.
[Switching to Thread 0x7fffec793700 (LWP 18526)]
pthread_sigmask (how=2, newmask=<optimized out>, oldmask=0x0) at ../nptl/sysdeps/pthread/pthread_sigmask.c:53
53	../nptl/sysdeps/pthread/pthread_sigmask.c: No such file or directory.
(gdb) c
Continuing.

Program received signal SIG34, Real-time event 34.
[Switching to Thread 0x7fffed795700 (LWP 18524)]
0x00007ffff3d416b7 in __GI_epoll_pwait (epfd=27, events=0x7fffed78e4d0, maxevents=128, timeout=-204204361, set=0x601000111010)
    at ../sysdeps/unix/sysv/linux/epoll_pwait.c:48
48	../sysdeps/unix/sysv/linux/epoll_pwait.c: No such file or directory.
(gdb) c
Continuing.

Program received signal SIG34, Real-time event 34.
[Switching to Thread 0x7fffecf94700 (LWP 18525)]
pthread_sigmask (how=2, newmask=<optimized out>, oldmask=0x0) at ../nptl/sysdeps/pthread/pthread_sigmask.c:53
53	../nptl/sysdeps/pthread/pthread_sigmask.c: No such file or directory.

```

It's very likely that I'm missing something, somewhere. But based on the available documentation and the logs that I'm reading I can't seem to figure it out, so any help is appreciated.

Lastly, the config I'm using:
```
cat /etc/scylla/scylla.yaml | grep -v '^#'


cluster_name: 'GeneralProd'

num_tokens: 256

data_file_directories:
    - /var/lib/scylla/data

commitlog_directory: /var/lib/scylla/commitlog

commitlog_sync: periodic
commitlog_sync_period_in_ms: 10000

commitlog_segment_size_in_mb: 32

seed_provider:
    # Addresses of hosts that are deemed contact points. 
    # Scylla nodes use this list of hosts to find each other and learn
    # the topology of the ring.  You must change this if you are running
    # multiple nodes!
    - class_name: org.apache.cassandra.locator.SimpleSeedProvider
      parameters:
          # seeds is actually a comma-delimited list of addresses.
          # Ex: "<ip1>,<ip2>,<ip3>"
          - seeds: "149.X.Y.115,149.X.Y.116,149.X.Y.117"

listen_address: 149.X.Y.132

native_transport_port: 9042

read_request_timeout_in_ms: 5000
write_request_timeout_in_ms: 2000


endpoint_snitch: GossipingPropertyFileSnitch

rpc_address: 149.X.Y.132
rpc_port: 9160

api_port: 10000
api_address: 127.0.0.1

batch_size_warn_threshold_in_kb: 5
```

Only the rpc_address and listen_address differ per host. Rest is identical.

**Comment by penberg** [comment]
So ``SIGUSR1``, ``SIG34``, and ``SIG35`` signales are used by the Seastar reactor so you should just "noprint" them in gdb.

Let's CC @asias on the streaming failure which seems to be the reason why your fifth node is unable to join.

**Comment by penberg** [comment]
@Freeaqingme, btw, to enable debug logging use the ``--default-log-level debug``  command line option (or even ``trace`` level).

**Comment by asias** [comment]
@Freeaqingme I believe it is the same bug we fixed for #1144 in scylla master.

https://github.com/scylladb/scylla/commit/45527fcffa6cf7aee80da3e37d3713ff3ce5f72f

We will backport it to 1.0 branch.

**Comment by Freeaqingme** [comment]
@asias I did review that report but did not immediately identify it as being the same issue. However, I just built f1cd52ff and after deploying it on all nodes I was able to join the new node right-away. Whatever the fix, it seems crucial for cluster operations.

As a sidenote, even though I did not time it it felt as if compiling took a really long time (even though I did not quantify it). I'm not sure if that's something that can be improved and if that warrants a separate issue.

TL;DR Works for me. Tnx!

**Comment by asias** [comment]
@Freeaqingme The fix is related to your issue because it fixes a hang in streaming operation which is needed to add a node. 

Yup, compiling scylla takes time ;-) It is due to the complexity of the code (c++14, template, etc.). We use distcc, ccache + faster machine to save compiling time.

Thanks for trying it out and confirming the fix.

**Comment by slivne** [comment]
closing the issue as it was confirmed the issue is resolved
